docs/async-await.md

```diff
@@ -0,0 +1,232 @@
+# Proposal: Async-await support
+
+## Introduction
+
+With the introduction of [async/await][SE-0296] in Swift 5.5, it is now possible to write asynchronous code without the need for callbacks. 
+
+Language support for [`AsyncSequence`][SE-0298] also allows for writing functions that return values over time.
+
+We would like to explore how we could offer APIs that make use of these new language features to allow users to run HTTPRequest using these new idioms.
+
+This proposal describes what these APIs could look like and explores some of the potential usability concerns.
+
+## Proposed API additions
+
+### New `HTTPClientRequest` type
+
+The proposed new `HTTPClientRequest` shall be a simple swift structure.
+
+```swift
+struct HTTPClientRequest {
+  /// The requests url.
+  var url: String
+  
+  /// The request's HTTPMethod
+  var method: HTTPMethod
+  
+  /// The request's headers
+  var headers: HTTPHeaders
```

Comment:
I was wondering whether `HTTPHeaders` is the right type here. For http/2 connections we'll need to normalise the header names to be lowercased and repackage them into `HPACKHeaders` which isn't cheap (in gRPC we used to use the http2-to-1 codecs and deal with http/1 types; we saw a pretty big increase in performance from using http2 directly, most of which came from not converting the headers). It'd be a shame to have to incur that cost here as well.

Keeping `HTTPHeaders` (over `HPACKHeaders` or a new AHC-provided headers type) is probably the preferred option at the moment since it's more or less a currency type but doesn't feel like an optimal solution.

Reply:
@ktoso @weissi @PeterAdams-A wdyt?

Reply:
In gRPC did you not just move the overhead to HTTP/1?  Ie, made HTTP/2 the default and now have to convert when on HTTP/1?

Reply:
Yes, exactly. It's a much clearer cut decision for gRPC since http/2 is _much_ more common; so much so that the cost of doing the extra work for http/1 doesn't really matter.

Reply:
Presumably at the point where this is currently created we don't know which HTTP version we'll be using?  Would we need to do some sort of delayed creation?  I think we need to provide and easy version to use, even if we also have a fast option.

Reply:
That's right, we need the headers before we know what version we're using. 

One option is AHC providing its own headers type which could wrap either an `HTTPHeaders` or `HPACKHeaders`.  This has the advantage of being able to change the backing type without breaking API if we decide to change our bias from http/1 to http/2 or vice versa. Of course this potentially makes things worse by adding a third headers type... 

Another would be making interop between `HTTPHeaders` and `HPACKHeaders` cheaper -- I'm not sure how possible this is though.

Reply:
Could make the third way a protocol with default implementations for both such that if we get the right one it's close to free, and wrong one is just the conversion we'd have to do anyway.